event_emitter/src/event.rs

```diff
@@ -0,0 +1,344 @@
+    collections::{btree_map::Entry, BTreeMap},
+};
+
+use iox_time::Time;
+
+use crate::measurement::TypedMeasurement;
+
+/// Value type for [`Event`] fields.
+#[derive(Debug, Clone)]
+pub enum FieldValue {
+    /// Signed integer.
+    I64(i64),
+
+    /// Unsigned integer.
+    U64(u64),
+
+    /// Float.
+    F64(f64),
+
+    /// Bool.
+    Bool(bool),
+
+    /// String.
+    String(Cow<'static, str>),
+}
+
+impl PartialEq for FieldValue {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::I64(l0), Self::I64(r0)) => l0 == r0,
+            (Self::U64(l0), Self::U64(r0)) => l0 == r0,
+            (Self::F64(l0), Self::F64(r0)) => l0 == r0,
+            (Self::Bool(l0), Self::Bool(r0)) => l0 == r0,
+            (Self::String(l0), Self::String(r0)) => l0 == r0,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for FieldValue {}
+
+impl From<i64> for FieldValue {
+    fn from(v: i64) -> Self {
+        Self::I64(v)
+    }
+}
+
+impl From<u64> for FieldValue {
+    fn from(v: u64) -> Self {
+        Self::U64(v)
+    }
+}
+
+impl From<f64> for FieldValue {
+    fn from(v: f64) -> Self {
+        Self::F64(v)
+    }
+}
+
+impl From<bool> for FieldValue {
+    fn from(v: bool) -> Self {
+        Self::Bool(v)
+    }
+}
+
+impl From<&'static str> for FieldValue {
+    fn from(v: &'static str) -> Self {
+        Self::String(v.into())
+    }
+}
+
+impl From<String> for FieldValue {
+    fn from(v: String) -> Self {
+        Self::String(v.into())
+    }
+}
+
+/// Typed InfluxDB-style event.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Event<M>
+where
+    M: Into<&'static str>,
+{
+    measurement: M,
+    time: Time,
+    tags: BTreeMap<&'static str, Cow<'static, str>>,
+    fields: BTreeMap<&'static str, FieldValue>,
+}
+
+impl<M> Event<M>
+where
+    M: Into<&'static str>,
+{
+    /// Create new measurement.
+    pub fn new(measurement: M, time: Time) -> Self {
+        Self {
+            measurement,
+            time,
+            tags: BTreeMap::default(),
+            fields: BTreeMap::default(),
+        }
+    }
+
+    /// Measurement.
+    pub fn measurement(&self) -> &M {
+        &self.measurement
+    }
+
+    /// Timestamp.
+    pub fn time(&self) -> Time {
+        self.time
+    }
+
+    /// Get all tags.
+    ///
+    /// The iterator elements are sorted by tag name.
+    pub fn tags(&self) -> impl Iterator<Item = (&'static str, &str)> {
+        self.tags.iter().map(|(k, v)| (*k, v.as_ref()))
+    }
+
+    /// Get all fields.
+    ///
+    /// The iterator elements are sorted by field name.
+    pub fn fields(&self) -> impl Iterator<Item = (&'static str, &FieldValue)> {
+        self.fields.iter().map(|(k, v)| (*k, v))
+    }
+
+    /// Add new tag.
+    ///
+    /// # Panic
+    /// Panics if a tag or a field with the same name already exists. Also panics if a tag called `"time"` is used.
+    pub fn add_tag_mut<V>(&mut self, name: &'static str, value: V) -> &mut Self
```

Comment:
```suggestion
    pub fn add_tag<V>(&mut self, name: &'static str, value: V)
```

I don't think this needs to be a builder-style method

Reply:
done